alibi_detect/cd/tensorflow/learnt_kernel.py

```diff
@@ -0,0 +1,208 @@
+        """
+        A module that wraps around the kernel. When passed a batch of reference and batch of test
+        instances it returns an estimate of a correlate of test power.
+        Equation 4 of https://arxiv.org/abs/2002.09116
+        """
+        def __init__(self, kernel: tf.keras.Model, var_reg: float):
+            super().__init__()
+            self.config = {'kernel': kernel, 'var_reg': var_reg}
+            self.kernel = kernel
+            self.var_reg = var_reg
+
+        def call(self, x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+            k_xx, k_yy, k_xy = self.kernel(x, x), self.kernel(y, y), self.kernel(x, y)
+            h_mat = k_xx + k_yy - k_xy - tf.transpose(k_xy)
+
+            mmd2_est = (tf.reduce_sum(h_mat)-tf.linalg.trace(h_mat))/(len(x)*(len(x)-1))
+            var_est = (4*tf.reduce_sum(tf.reduce_sum(h_mat, axis=-1)**2)/(len(x)**3) -
+                       4*tf.reduce_sum(h_mat)**2/(len(x)**4))
+            reg_var_est = var_est + self.var_reg
+
+            return mmd2_est/tf.math.sqrt(reg_var_est)
+
+    def score(self, x: np.ndarray) -> Tuple[float, float, np.ndarray]:
+        """
+        Compute the p-value resulting from a permutation test using the maximum mean discrepancy
+        as a distance measure between the reference data and the data to be tested. The kernel
+        used within the MMD is first trained to maximise an estimate of the resulting test power.
+
+        Parameters
+        ----------
+        x
+            Batch of instances.
+
+        Returns
+        -------
+        p-value obtained from the permutation test, the MMD^2 between the reference and test set
+        and the MMD^2 values from the permutation test.
+        """
+        x_ref, x_cur = self.preprocess(x)
+        (x_ref_tr, x_cur_tr), (x_ref_te, x_cur_te) = self.get_splits(x_ref, x_cur)
+        ds_ref_tr, ds_cur_tr = self.dataset(x_ref_tr), self.dataset(x_cur_tr)
+        
+        self.kernel = clone_model(self.original_kernel) if self.retrain_from_scratch else self.kernel
+        train_args = [self.j_hat, (ds_ref_tr, ds_cur_tr)]
+        LearntKernelDriftTF.trainer(*train_args, **self.train_kwargs)  # type: ignore
+
+        kernel_mat = self.kernel_mat_fn(x_ref_te, x_cur_te, self.kernel)
+        kernel_mat = kernel_mat - tf.linalg.diag(tf.linalg.diag_part(kernel_mat))  # zero diagonal
+        mmd2 = mmd2_from_kernel_matrix(kernel_mat, len(x_cur_te), permute=False, zero_diag=False).numpy()
+        mmd2_permuted = np.array(
+            [mmd2_from_kernel_matrix(kernel_mat, len(x_cur_te), permute=True, zero_diag=False).numpy()
+                for _ in range(self.n_permutations)]
+        )
+        p_val = (mmd2 <= mmd2_permuted).mean()
+        return p_val, mmd2, mmd2_permuted
+
+    @staticmethod
+    def trainer(
+        j_hat: JHat,
+        datasets: Tuple[tf.keras.utils.Sequence, tf.keras.utils.Sequence],
+        optimizer: tf.keras.optimizers = tf.keras.optimizers.Adam,
```

Comment:
The default optimizer requires an initialized optimizer, e.g. `tf.keras.optimizers.Adam(learning_rate=1e-3)`.

Reply:
Ok I'll change the default I think as I'm worried about having a single optimizer with continued state between separate calls to predict.